### PR TITLE
Show normalized version (incl metadata) in edit/delete package dropdown.

### LINKED ISSUE #4756: Edit experience is broken for SemVer 2.0.0 packages with build metadata
Try to edit a package version that has Semantic Versioning 2.0.0 due to build metadata. The dropdown of the package versions do not show the build metadata information.

Example: Version 1.2.4 is a SemVer v2.0.0 package with full version as 1.2.4+buildmeta but I do not see that in the dropdown:
image

If you choose to edit any such SemVer 2.0.0 package in the drop down, it goes to an 404 page:
image

### PR DESCRIPTION
**Pull request by xavierdecoster**
Fixes #4756 

Show normalized version (incl metadata) in edit package dropdown.
Fix semver2 package edit navigation (URLs must always use normalized version).

### CODE REVIEW

File: src/NuGetGallery/Views/Packages/Edit.cshtml [hunk 1]
```diff
@@ -39,8 +41,8 @@
                                 @Html.DropDownList("version-selection",
                             new SelectList(Model.PackageVersions.Select(e => new
                             {
-                                text = e.NormalizedVersion + (e.IsLatestSemVer2 ? " (Latest)" : ""),
```
Comment (joelverhagen) [inline src/NuGetGallery/Views/Packages/Edit.cshtml]: Can this be done in the `ViewModel` or `Controller` instead? Unit testable.

Comment (xavierdecoster) [reply src/NuGetGallery/Views/Packages/Edit.cshtml]: Will have to happen in the `Controller` then as it relies on `UrlHelper`.

Comment (xavierdecoster) [reply src/NuGetGallery/Views/Packages/Edit.cshtml]: Updated.

File: src/NuGetGallery/Views/Packages/Edit.cshtml [hunk 2]
```diff
@@ -36,12 +36,7 @@
                         <div class="row">
                             <label for="input-select-version" id="package-version-label" class="col-xs-12">
                                 <span>Select a package version</span>
-                                @Html.DropDownList("version-selection",
-                            new SelectList(Model.PackageVersions.Select(e => new
-                            {
-                                text = e.NormalizedVersion + (e.IsLatestSemVer2 ? " (Latest)" : ""),
-                                url = Url.EditPackage(Model.PackageId, e.Version)
-                            }), "url", "text", Url.EditPackage(Model.PackageId, Model.Version)),
+                                @Html.DropDownList("version-selection", Model.VersionSelectList,
```
Comment (chenriksson) [inline src/NuGetGallery/Views/Packages/Edit.cshtml]: There's similar code in Delete.cshtml, does that view have the same issue?

Seems like Delete should be consistent with Edit (i.e., moving SelectList to the model), but not sure we can given that DeletePackageViewModel descends from DisplayPackageViewModel. This relation seems a bit strange... wondering if Delete should be separated from DisplayPackage.

File: tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs [hunk 3]
```diff
@@ -25,6 +25,7 @@
 using NuGetGallery.Security;
 using NuGetGallery.TestUtils;
 using Xunit;
+using NuGet.Versioning;
```
Comment (chenriksson) [inline tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs]: nit: sort

Comment (xavierdecoster) [reply tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs]: Fixed in https://github.com/NuGet/NuGetGallery/pull/4785/commits/1e5043d38ca246504b6e49edaa8892c6d1dcfcf1

File: src/NuGetGallery/Views/Packages/Edit.cshtml (hunk 2 above)
Comment (xavierdecoster) [reply src/NuGetGallery/Views/Packages/Edit.cshtml]: Good catch! 
Fixed in https://github.com/NuGet/NuGetGallery/pull/4785/commits/1e5043d38ca246504b6e49edaa8892c6d1dcfcf1

File: src/NuGetGallery/Controllers/PackagesController.cs [hunk 4]
```diff
@@ -922,6 +921,16 @@ public virtual ActionResult Delete(string id, string version)
             }
 
             var model = new DeletePackageViewModel(package, ReportMyPackageReasons);
+
+            model.VersionSelectList = new SelectList(
+                model.PackageVersions
+                .Where(p => p.Deleted == false)
```
Comment (joelverhagen) [inline src/NuGetGallery/Controllers/PackagesController.cs]: `p.PackageStatusKey == PackageStatus.Available`

Comment (joelverhagen) [reply src/NuGetGallery/Controllers/PackagesController.cs]: Nevermind. This is on the view model, not `Package`. 💤

File: src/NuGetGallery/ViewModels/DeletePackageViewModel.cs [hunk 5]
```diff
@@ -1,10 +1,11 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using NuGet.Versioning;
 using System.Collections.Generic;
 using System.Globalization;
 using System.Linq;
-using NuGet.Versioning;
+using System.Web.Mvc;
```
Comment (chenriksson) [inline src/NuGetGallery/ViewModels/DeletePackageViewModel.cs]: nit: sort usings

File: tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs [hunk 6]
```diff
@@ -1212,14 +1211,101 @@ public class TheEditMethod
                 Assert.Equal(@"~\Bar.cshtml", ((RedirectResult)result).Url);
             }
 
+            [Fact]
+            public async Task UsesNormalizedVersionsInUrlsInSelectList()
+            {
```
Comment (chenriksson) [inline tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs]: Should there be a test for Delete as well?

Comment (xavierdecoster) [reply tests/NuGetGallery.Facts/Controllers/PackagesControllerFacts.cs]: Added in https://github.com/NuGet/NuGetGallery/pull/4785/commits/115532de6391df3b40224e1758a72a03e4d94e28

**Comment by xavierdecoster** [comment]
🔔

**Pull request merged**